File: code/py/src/modular_api/graphql/read/sql_read_contract.py
"""Pure GraphQL SQL read contract for Stage 5."""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Iterable, Mapping


class SqlReadCommandPurpose(str, Enum):
    ITEM = "item"
    COLLECTION = "collection"
    RELATION_BATCH = "relation-batch"
    COUNT = "count"


@dataclass(frozen=True, slots=True)
class SqlParameter:
    name: str
    value: Any
    type: str | None = None


@dataclass(frozen=True, slots=True)
class SqlReadCommand:
    engine: str
    sql: str
    parameters: tuple[SqlParameter, ...]
    purpose: SqlReadCommandPurpose


@dataclass(frozen=True, slots=True)
class ReadExecutionContext:
    request_id: str | None = None
    principal: Any = None
    tenant_id: str | None = None
    telemetry: Any = None


@dataclass(frozen=True, slots=True)
class RowSet:
    rows: tuple[dict[str, Any], ...]
    row_count: int

    @staticmethod
    def normalize(raw_rows: Iterable[Mapping[object, Any]]) -> RowSet:
        rows: list[dict[str, Any]] = []
        for raw_row in raw_rows:
            row = {str(key): raw_row[key] for key in sorted(raw_row, key=lambda value: str(value))}
            rows.append(row)
        return RowSet(rows=tuple(rows), row_count=len(rows))


class ReadExecutor(ABC):
    @abstractmethod
    async def execute(self, command: SqlReadCommand, context: ReadExecutionContext) -> RowSet:
        raise NotImplementedError

    async def close(self) -> None:
        return None


class SqlFilterOperator(str, Enum):
    EQ = "eq"
    NE = "ne"
    IN_LIST = "inList"
    LT = "lt"
    LTE = "lte"
    GT = "gt"
    GTE = "gte"
    IS_NULL = "isNull"
    CONTAINS = "contains"
    STARTS_WITH = "startsWith"
    ENDS_WITH = "endsWith"


class SqlFilterNode:
    pass


@dataclass(frozen=True, slots=True)
class SqlFilterCondition(SqlFilterNode):
    field: str
    operator: SqlFilterOperator
    value: Any


class SqlFilterGroupKind(str, Enum):
    AND = "and"
    OR = "or"
    NOT = "not"


@dataclass(frozen=True, slots=True)
class SqlFilterGroup(SqlFilterNode):
    kind: SqlFilterGroupKind
    nodes: tuple[SqlFilterNode, ...]

    @staticmethod
    def and_(nodes: tuple[SqlFilterNode, ...]) -> SqlFilterGroup:
        return SqlFilterGroup(kind=SqlFilterGroupKind.AND, nodes=nodes)

    @staticmethod
    def or_(nodes: tuple[SqlFilterNode, ...]) -> SqlFilterGroup:
        return SqlFilterGroup(kind=SqlFilterGroupKind.OR, nodes=nodes)

    @staticmethod
    def not_(node: SqlFilterNode) -> SqlFilterGroup:
        return SqlFilterGroup(kind=SqlFilterGroupKind.NOT, nodes=(node,))


class SqlSortDirection(str, Enum):
    ASC = "asc"
    DESC = "desc"


@dataclass(frozen=True, slots=True)
class SqlOrderByClause:
    field: str
    direction: SqlSortDirection


@dataclass(frozen=True, slots=True)
class SqlPage:
    limit: int
    offset: int


@dataclass(frozen=True, slots=True)
class SqlItemSelection:
    object_id: str
    projected_fields: tuple[str, ...]
    key: dict[str, Any]


@dataclass(frozen=True, slots=True)
class SqlCollectionSelection:
    object_id: str
    projected_fields: tuple[str, ...]
    filter: SqlFilterNode | None = None
    order_by: tuple[SqlOrderByClause, ...] = ()
    page: SqlPage | None = None


@dataclass(frozen=True, slots=True)
class SqlCountSelection:
    object_id: str
    filter: SqlFilterNode | None = None


@dataclass(frozen=True, slots=True)
class SqlRelationBatchSelection:
    source_object_id: str
    relation_name: str
    projected_fields: tuple[str, ...]
    parent_keys: tuple[dict[str, Any], ...]

File: code/dart/modular_api/lib/src/graphql/read/sql_read_contract.dart
import 'dart:collection';

enum SqlReadCommandPurpose {
  item,
  collection,
  relationBatch,
  count,
}

final class SqlParameter {
  const SqlParameter({
    required this.name,
    required this.value,
    this.type,
  });

  final String name;
  final String? type;
  final Object? value;
}

final class SqlReadCommand {
  const SqlReadCommand({
    required this.engine,
    required this.sql,
    required this.parameters,
    required this.purpose,
  });

  final String engine;
  final String sql;
  final List<SqlParameter> parameters;
  final SqlReadCommandPurpose purpose;
}

final class ReadExecutionContext {
  const ReadExecutionContext({
    this.requestId,
    this.principal,
    this.tenantId,
    this.telemetry,
  });

  final String? requestId;
  final Object? principal;
  final String? tenantId;
  final Object? telemetry;
}

final class RowSet {
  const RowSet({required this.rows, required this.rowCount});

  final List<Map<String, Object?>> rows;
  final int rowCount;

  static RowSet normalize(Iterable<Map<Object?, Object?>> rawRows) {
    final rows = rawRows.map((rawRow) {
      final sortedEntries = rawRow.entries
          .map((entry) => MapEntry(entry.key.toString(), entry.value))
          .toList(growable: false)
        ..sort((left, right) => left.key.compareTo(right.key));
      final row = SplayTreeMap<String, Object?>();
      for (final entry in sortedEntries) {
        row[entry.key] = entry.value;
      }
      return Map<String, Object?>.unmodifiable(row);
    }).toList(growable: false);

    return RowSet(rows: rows, rowCount: rows.length);
  }
}

abstract class SqlReadExecutor {
  Future<RowSet> execute(SqlReadCommand command, ReadExecutionContext context);

  Future<void> close() async {}
}

enum SqlFilterOperator {
  eq,
  ne,
  inList,
  lt,
  lte,
  gt,
  gte,
  isNull,
  contains,
  startsWith,
  endsWith,
}

abstract class SqlFilterNode {
  const SqlFilterNode();
}

final class SqlFilterCondition extends SqlFilterNode {
  const SqlFilterCondition({
    required this.field,
    required this.operator,
    required this.value,
  });

  final String field;
  final SqlFilterOperator operator;
  final Object? value;
}

enum SqlFilterGroupKind {
  and,
  or,
  not,
}

final class SqlFilterGroup extends SqlFilterNode {
  const SqlFilterGroup._({required this.kind, required this.nodes});

  const SqlFilterGroup.and(List<SqlFilterNode> nodes)
      : this._(kind: SqlFilterGroupKind.and, nodes: nodes);

  const SqlFilterGroup.or(List<SqlFilterNode> nodes)
      : this._(kind: SqlFilterGroupKind.or, nodes: nodes);

  SqlFilterGroup.not(SqlFilterNode node)
      : kind = SqlFilterGroupKind.not,
        nodes = <SqlFilterNode>[node];

  final SqlFilterGroupKind kind;
  final List<SqlFilterNode> nodes;
}

enum SqlSortDirection {
  asc,
  desc,
}

final class SqlOrderByClause {
  const SqlOrderByClause({required this.field, required this.direction});

  final String field;
  final SqlSortDirection direction;
}

final class SqlPage {
  const SqlPage({required this.limit, required this.offset});

  final int limit;
  final int offset;
}

final class SqlItemSelection {
  const SqlItemSelection({
    required this.objectId,
    required this.projectedFields,
    required this.key,
  });

  final String objectId;
  final List<String> projectedFields;
  final Map<String, Object?> key;
}

final class SqlCollectionSelection {
  const SqlCollectionSelection({
    required this.objectId,
    required this.projectedFields,
    this.filter,
    this.orderBy = const <SqlOrderByClause>[],
    this.page,
  });

  final String objectId;
  final List<String> projectedFields;
  final SqlFilterNode? filter;
  final List<SqlOrderByClause> orderBy;
  final SqlPage? page;
}

final class SqlCountSelection {
  const SqlCountSelection({required this.objectId, this.filter});

  final String objectId;
  final SqlFilterNode? filter;
}

final class SqlRelationBatchSelection {
  const SqlRelationBatchSelection({
    required this.sourceObjectId,
    required this.relationName,
    required this.projectedFields,
    required this.parentKeys,
  });

  final String sourceObjectId;
  final String relationName;
  final List<String> projectedFields;
  final List<Map<String, Object?>> parentKeys;
}